pkg/batcher/transport_message_batcher.go

```diff
@@ -0,0 +1,362 @@
+	dst := b.buffer[b.bufferIndex:]
+	start := 0
+	// group
+	start += writeType(dst[start:], message_group_codec_type)
+	msgGroupLenIndex := start
+	start += writeLen(dst[start:], 0)
+
+	// list
+	start += writeType(dst[start:], message_list_codec_type)
+	msgListLenIndex := start
+	start += writeLen(dst[start:], 0)
+
+	// msg
+	start += writeType(dst[start:], raw_message_codec_type)
+	msgLenIndex := start
+	start += writeLen(dst[start:], 0)
+
+	// msg id
+	start += writeType(dst[start:], message_id_codec_type)
+	msgIdLenIndex := start
+	start += writeLen(dst[start:], 0)
+
+	start += writeString(dst[start:], session_alias_codec_type, alias)
+	start += writeDirection(dst[start:], direction)
+	sequence := b.seqProvider.nextSeq(alias, direction)
+	start += writeLongValue(dst[start:], sequence_codec_type, sequence)
+	start += writeLongCollection(dst[start:], subsequence_codec_type, nil)
+	start += writeTime(dst[start:], time.Now().UnixNano())
+	writeLen(dst[msgIdLenIndex:], start-msgIdLenIndex-length_size)
+
+	// metadata
+	start += writeMetadata(dst[start:], metadata)
+
+	// protocol
+	start += writeProtocol(dst[start:], orDefaultIfEmpty(protocol, b.protocol))
+
+	// body
+	start += writeType(dst[start:], raw_body_codec_type)
+	start += writeLen(dst[start:], len(data))
+	start += copy(dst[start:], data)
+
+	writeLen(dst[msgLenIndex:], start-msgLenIndex-length_size)
+	writeLen(dst[msgListLenIndex:], start-msgListLenIndex-length_size)
+	writeLen(dst[msgGroupLenIndex:], start-msgGroupLenIndex-length_size)
+
+	b.bufferIndex += start
+
+	b.groupIndex++
+}
+
+func (b *messageBatcher) flush() {
+	if b.groupIndex == 0 {
+		log.Trace().Msg("Flushing has skipped because buffer is empty")
+		return
+	}
+	log.Trace().Int("messageCount", b.groupIndex).Msg("Store messages")
+	batchLen := b.bufferIndex - b.batchLenIndex - length_size
+	messagesLen := b.bufferIndex - b.groupsLenIndex - length_size
+	writeLen(b.buffer[b.batchLenIndex:], batchLen)
+	writeLen(b.buffer[b.groupsLenIndex:], messagesLen)
+
+	if err := b.router.SendRawAll(b.buffer[:b.bufferIndex], th2_pin_attribute); err != nil {
+		log.Panic().Err(err).Msg("flushing message failure")
+	}
+
+	b.reset()
+}
+
+func (b *messageBatcher) reset() {
+	log.Trace().Int("bufferIndex", b.bufferIndex).Msg("Resetting state")
+
+	b.groupIndex = 0
+	b.batchLenIndex = 0
+	b.groupsLenIndex = 0
+	b.bufferIndex = 0
+	b.bufferIndex += writeType(b.buffer[b.bufferIndex:], group_batch_codec_type)
+	b.batchLenIndex = b.bufferIndex
+	b.bufferIndex += writeLen(b.buffer[b.bufferIndex:], 0)
+
+	b.bufferIndex += writeBook(b.buffer[b.bufferIndex:], b.book)
+	b.bufferIndex += writeGroup(b.buffer[b.bufferIndex:], b.group)
+
+	b.bufferIndex += writeType(b.buffer[b.bufferIndex:], group_list_codec_type)
+	b.groupsLenIndex = b.bufferIndex
+	b.bufferIndex += writeLen(b.buffer[b.bufferIndex:], 0)
+}
+
+func (b *messageBatcher) computeOverhead(data []byte, args MessageArguments) int {
```

Comment:
note: looks like the `metadata` size (message properties) is not taken into account here. Could you please take a look? Maybe I am wrong here

Reply:
Yes you are right. I will look at reusing https://github.com/th2-net/transport-go